tests/end_to_end/end_to_end.cpp

```diff
@@ -1064,4 +1078,347 @@ TEST_CASE("link_tests", "[end_to_end]")
     REQUIRE(hook.attach(program) == EBPF_INVALID_ARGUMENT);
 
     hook.detach();
-}
\ No newline at end of file
+}
+
+TEST_CASE("map_reuse", "[end_to_end]")
+{
+    _test_helper_end_to_end test_helper;
+    single_instance_hook_t hook(EBPF_PROGRAM_TYPE_XDP, EBPF_ATTACH_TYPE_XDP);
+    program_info_provider_t xdp_program_info(EBPF_PROGRAM_TYPE_XDP);
+
+    // First create and pin the maps manually.
+    int inner_map_fd = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(inner_map_fd > 0);
+
+    int outer_map_fd = bpf_create_map_in_map(BPF_MAP_TYPE_HASH_OF_MAPS, nullptr, sizeof(__u32), inner_map_fd, 1, 0);
+    REQUIRE(outer_map_fd > 0);
+
+    // Verify we can insert the inner map into the outer map.
+    __u32 outer_key = 0;
+    int error = bpf_map_update_elem(outer_map_fd, &outer_key, &inner_map_fd, 0);
+    REQUIRE(error == 0);
+
+    // Pin the outer map
+    error = bpf_obj_pin(outer_map_fd, "/ebpf/global/outer_map");
+    REQUIRE(error == 0);
+
+    int port_map_fd = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(port_map_fd > 0);
+
+    // Pin port map
+    error = bpf_obj_pin(port_map_fd, "/ebpf/global/port_map");
+    REQUIRE(error == 0);
+
+    // Add an entry in the inner map.
+    __u32 key = 0;
+    __u32 value = 200;
+    error = bpf_map_update_elem(inner_map_fd, &key, &value, BPF_ANY);
+    REQUIRE(error == 0);
+
+    program_load_attach_helper_t program_helper(
+        SAMPLE_PATH "map_reuse.o", EBPF_PROGRAM_TYPE_XDP, "lookup_update", EBPF_EXECUTION_ANY, hook);
+
+    auto packet = prepare_udp_packet(10, ETHERNET_TYPE_IPV4);
+    xdp_md_t ctx{packet.data(), packet.data() + packet.size()};
+    int hook_result;
+
+    REQUIRE(hook.fire(&ctx, &hook_result) == EBPF_SUCCESS);
+    REQUIRE(hook_result == 200);
+
+    key = 0;
+    __u32 port_map_value;
+    REQUIRE(bpf_map_lookup_elem(port_map_fd, &key, &port_map_value) == EBPF_SUCCESS);
+    REQUIRE(port_map_value == 200);
+
+    REQUIRE(_get_total_map_count() == 4);
+
+    Platform::_close(outer_map_fd);
+    Platform::_close(inner_map_fd);
+    Platform::_close(port_map_fd);
+
+    REQUIRE(ebpf_object_unpin("/ebpf/global/outer_map") == EBPF_SUCCESS);
+    REQUIRE(ebpf_object_unpin("/ebpf/global/port_map") == EBPF_SUCCESS);
+}
```

Comment:
recommend verifying that all maps have been cleaned up after the unpins, to make sure there's no refcount leak.  Same for the other tests.

```suggestion

    // Verify that all maps were successfully removed.
    uint32_t id;
    REQUIRE(bpf_map_get_next_id(0, &id) < 0);
    REQUIRE(errno == ENOENT);
}
```

Or even better, add this to the _test_helper_end_to_end::~_test_helper_end_to_end() destructor.

Reply:
Done.